march_hardware_interface/src/march_hardware_interface.cpp

```diff
@@ -448,15 +448,21 @@ void MarchHardwareInterface::iMotionCubeStateCheck(size_t joint_index)
   if (imc_state.state == march::IMCState::FAULT)
   {
     this->march_robot_->stopEtherCAT();
-    std::ostringstream error_stream;
-    error_stream << "IMotionCube of joint " << joint_names_[joint_index].c_str() << " is in fault state "
-                 << imc_state.state.getString() << std::endl;
-    error_stream << "Detailed Error: " << imc_state.detailedErrorDescription << "(" << imc_state.detailedError << ")"
-                 << std::endl;
-    error_stream << "Motion Error: " << imc_state.motionErrorDescription << "(" << imc_state.motionError << ")"
-                 << std::endl;
-
-    throw std::runtime_error(error_stream.str());
+      for (int i = 0; i<8; i++){
+          march::IMotionCubeState imc_state = march_robot_->getJoint(joint_names_[joint_index]).getIMotionCubeState();
+          if (imc_state.state == march::IMCState::FAULT){
+              std::ostringstream error_stream;
+              error_stream << "IMotionCube of joint " << joint_names_[joint_index].c_str() << " is in fault state "
+                           << imc_state.state.getString() << std::endl;
+              error_stream << "Detailed Error: " << imc_state.detailedErrorDescription << "(" << imc_state.detailedError << ")"
+                           << std::endl;
+              error_stream << "Motion Error: " << imc_state.motionErrorDescription << "(" << imc_state.motionError << ")"
+                           << std::endl;
+
+              throw std::runtime_error(error_stream.str());
```

Comment:
this command throws an error and shuts down the exoskeleton. You should not throw it again for every imc that is in fault state, but rather only once at the end if any imc has faulted. 

You can construct the error message by declaring `error_stream` before the loop, and then adding text on every iteration.

Reply:
So `std::ostringstream error_stream;` needs to be before the loop and `throw std::runtime_error(error_stream.str());` after the loop?

Reply:
I would change the std::runtime_error with a logging statement and place the throw after the loop.